### ISSUE TITLE
The execution result of enum variable in `==` is incorrect in vim9script

### ISSUE THREAD

**Issue by jiangyinzuo** [vim/vim#15067]
### Steps to reproduce

In x1.vim
```vim
vim9script

enum Way
  Prompt,
  Terminal
endenum

var way: Way

if false
  way = Way.Prompt
  echom 'way: prompt'
elseif true
  way = Way.Terminal
  echom 'way: terminal'
else
  way = Way.Prompt
  echom 'way: prompt'
endif

if way == Way.Prompt
  echom 'way: prompt'
else
  echom 'way: terminal'
endif
```

In x2.vim
```vim
vim9script

var way: number

if false
  way = 0
  echom 'way: 0'
elseif true
  way = 1
  echom 'way: 1'
else
  way = 0
  echom 'way: 0'
endif

if way == 0
  echom 'way: 0'
else
  echom 'way: 1'
endif
```

Run `vim -U NONE --noplugin -S x1.vim` , output
```
way: terminal
way: prompt
```

Run `vim -U NONE --noplugin -S x2.vim`, output
```
way: 1
way: 1
```

In x3.vim
```vim
vim9script

enum Way
  Prompt,
  Terminal
endenum

export var way: Way

def g:Foo()
  g:way = Way.Terminal
  if g:way == Way.Prompt
    echom 'way: prompt'
  else
    echom 'way: terminal'
  endif
enddef

g:Foo()
echo g:way
```

output
```
way: prompt
enum Way.Terminal {name: Terminal, ordinal: 1}
```

### Expected behaviour

For x1.vim, output
```
way: terminal
way: terminal
```

For x2.vim, output
```
way: 1
way: 1
```

For x3.vim, output
```
way: terminal
enum Way.Terminal {name: Terminal, ordinal: 1}
```

### Version of Vim

9.1.0512

### Environment

Ubuntu 22.04

**Comment by jiangyinzuo** [comment]
`==` seems buggy, but `is` seems correct.

```vim
vim9script

enum Way
  Prompt,
  Terminal
endenum

export var way: Way

def g:Foo()
  g:way = Way.Terminal
  if g:way is Way.Prompt
    echom 'way: prompt'
  else
    echom 'way: terminal'
  endif
enddef

g:Foo()
echo g:way
```

output
```
way: terminal
enum Way.Terminal {name: Terminal, ordinal: 1}
```

**Comment by zzzyxwvut** [comment]
Curiously, when the example files are opened and _then_  
sourced you do get the `is`-like results, as opposed to what  
you see with `vim --clean --cmd 'so x1.vim | q'` etc.

(Tested in v9.1.0507.)

**Comment by errael** [comment]
> Curiously, when the example files are opened and _then_
> sourced

Beat me to it. I just tried `vim -U NONE --noplugin -c :source /tmp/x1.vim` and see
```
way: terminal
way: terminal
```

What does the `-S` do?

**Comment by jiangyinzuo** [comment]
@zzzyxwvut  @errael  I left a comment to explain the interesting and strange behavior https://github.com/vim/vim/pull/15070#issuecomment-2183105133

**Comment by zzzyxwvut** [comment]
Also, an `is`-like result can be obtained for `x1.vim` without  
opening `x1.vim` but rather importing it, e.g. with `test.vim`

```
vim9script

import './x1.vim'
```

```
vim --clean test.vim

:so %
```

HOWEVER, sourcing does not produce an `is`-like result:

1)
```
vim --clean

:so x1.vim
```

2)
```
vim --clean

:so test.vim
```

And neither these commands:

```shell
vim --clean --cmd 'e x1.vim | so % | q'
vim --clean --cmd 'e test.vim | so % | q'
```

**Comment by jiangyinzuo** [comment]
https://github.com/vim/vim/pull/15076 fixes this issue, now #15070 is just a code clean up.
